### ISSUE TITLE
Domain contexts for settings supporting URL patterns

### ISSUE THREAD

**Issue by bzbug** [qutebrowser/qutebrowser#6922]
Four contexts:

1. Current domain
2. Current full URL
3. Sub-domains of the current domain
4. Sub-domains of the 2LD in the current domain (for example, if the current domain is `cdn.example.com`, the context would be `*.example.com`, not `*.cdn.example.com`).

... and four hotkeys, one for each. The user enters the hotkey of the context - and he automatically enters cmd mode for that chosen context and with `:set` already in the cmdline. And then the user can use the existing system of completions for his chosen context - just like in global completion of `set` command, the user can see what settings are available to him (only per-domain, so there will be less), and, more importantly, what values they already have for this chosen context, and he can change them on the fly.

Would be nice if it's also possible, when the context is chosen and the browser is in cmd mode, to display the context before the `:set` in the form of the actual (sub-)domain or a word `page` for the whole current URL (since it's supposed to be already displayed in the statusbar).

`config-diff` is nice, but it's not very convenient for a look up, especially since it doesn't have domain contexts. And copying current domain and then altering it by hand to include sub-domains is not very quick either, especially considering that changing permissions on the fly for the current URL, domain or its sub-domains seems like something that should not require manual handling of the domains. The existing hotkeys with value cycling (e.g. for cookies) don't mean there shouldn't be a 'proper' _and_ convenient way to change the settings per-context, and value cycling lacks info on currently set values for chosen contexts.

**Comment by The-Compiler** [comment]
I don't really think it makes sense to have this as a global context of some sorts, given that it only affects `:set` (and a couple of closely related `:config-*` commands perhaps). Instead, the `:set` completion should take into account that `-u` is given, and show values/settings for that pattern instead. I thought we had an issue about completions taking into account command flags already, but I can't seem to find it. cc @rcorre

Then the only missing bit to bind keys to something like `:set-cmd-text -s set -u {url:host}`, similar to how `ss` is already bound. As for 4., we would first need #756, but that could just be another `{url:...}` replacement then.

**Comment by toofar** [comment]
> I thought we had an issue about completions taking into account command flags already

Same, I thought I even commented on it once... Maybe it got merged into https://github.com/qutebrowser/qutebrowser/issues/32 at some point (maybe https://github.com/qutebrowser/qutebrowser/issues/1532). I think there was discussion of it most recently on https://github.com/qutebrowser/qutebrowser/issues/5537

**Comment by toofar** [comment]
Sorry if I'm not 100% following this but I will try to summarize. The original proposal was to do some kind of modal / domain context thing. But that would be a whole new piece of UI and state so we would like to try to tackle this with the existing completion stuff. So I think there are four parts:

1. support more common URL pattern substitutions (after #756)
2. add completion for common URL patterns based on the current tab's url for when the cursor is like this `:set -u |`: should make it easier to set a bunch of URL specific settings with less mistakes or having to remember rarely used syntax or mnemonics (I find the existing ones hard enough to remember). Probably tracked in #32 somewhere already? Maybe "Choices for flags with argument", it would require the same (new) infrastructure as that one would anyway
3. filter the list of settings once you have a URL pattern in there, for when the cursor is like this `:set -u some.site |`: will filter the list down to just settings that support URL patterns, hopefully that isn't more confusing with not being able to find a setting you are looking for
4. add a new category to the `:set` completion to show settings that already have per-URL settings with the same (or similar) patterns to the one you have entered: should make it easier to see what settings you already have set for the URL pattern you have entered. This would look similar to the `:bind` completion where there are separate categories for bindings with the same keys and for all the commands you can bind (try `:bind <Esc> |` if you haven't noticed this before).

If that sounds like a plan I would like to update the issue title to something more generic like "Better completion to support for per-url settings." so that it is clearer to me what it is about and seems (a little bit) more actionable.
